server/sonar-server/src/main/java/org/sonar/server/authentication/InitFilter.java

```diff
@@ -63,35 +70,80 @@ public UrlPattern doGetPattern() {
+      String key = requestUri.replace(context, "");
+      if (!isNullOrEmpty(key)) {
+        return key;
+      }
+    }
+    return null;
+  }
+
+  private void handleProvider(HttpServletRequest request, HttpServletResponse response, IdentityProvider provider) {
     try {
-      keyProvider = extractKeyProvider(requestURI, server.getContextPath() + INIT_CONTEXT);
-      IdentityProvider provider = identityProviderRepository.getEnabledByKey(keyProvider);
       if (provider instanceof BaseIdentityProvider) {
-        BaseIdentityProvider baseIdentityProvider = (BaseIdentityProvider) provider;
-        baseIdentityProvider.init(baseContextFactory.newContext(httpRequest, (HttpServletResponse) response, baseIdentityProvider));
+        handleBaseIdentityProvider(request, response, (BaseIdentityProvider) provider);
       } else if (provider instanceof OAuth2IdentityProvider) {
-        OAuth2IdentityProvider oAuth2IdentityProvider = (OAuth2IdentityProvider) provider;
-        oAuth2IdentityProvider.init(oAuth2ContextFactory.newContext(httpRequest, (HttpServletResponse) response, oAuth2IdentityProvider));
+        handleOAuth2IdentityProvider(request, response, (OAuth2IdentityProvider) provider);
       } else {
-        throw new UnsupportedOperationException(format("Unsupported IdentityProvider class: %s ", provider.getClass()));
+        handleError(response, format("Unsupported IdentityProvider class: %s", provider.getClass()));
       }
-    } catch (UnauthorizedException e) {
-      handleUnauthorizedError(e, (HttpServletResponse) response);
+    } catch (AuthenticationException e) {
+      authenticationEvent.failure(request, e);
+      handleAuthenticationError(e, response);
     } catch (Exception e) {
-      handleError(e, (HttpServletResponse) response, format("Fail to initialize authentication with provider '%s'", keyProvider));
+      handleError(e, response, format("Fail to initialize authentication with provider '%s'", provider.getKey()));
     }
   }
 
-  public static String extractKeyProvider(String requestUri, String context) {
-    if (requestUri.contains(context)) {
-      String key = requestUri.replace(context, "");
-      if (!isNullOrEmpty(key)) {
-        return key;
-      }
+  private void handleBaseIdentityProvider(HttpServletRequest request, HttpServletResponse response, BaseIdentityProvider provider) {
+    try {
+      provider.init(baseContextFactory.newContext(request, response, provider));
+    } catch (UnauthorizedException e) {
+      throw AuthenticationException.newBuilder()
+        .setSource(Source.external(provider))
+        .setMessage(e.getMessage())
+        .setPublicMessage(e.getMessage())
+        .build();
+    }
+  }
+
+  private void handleOAuth2IdentityProvider(HttpServletRequest request, HttpServletResponse response, OAuth2IdentityProvider provider) {
+    try {
+      provider.init(oAuth2ContextFactory.newContext(request, response, provider));
+    } catch (UnauthorizedException e) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Either log or rethrow this exception. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1166)